src/c-unchop.c

```diff
@@ -82,88 +82,74 @@ static SEXP vec_unchop(SEXP x,
 
   // `out_size` is computed from `indices`
   for (R_len_t i = 0; i < x_size; ++i) {
-    SEXP elt = VECTOR_ELT(x, i);
+    SEXP x = VECTOR_ELT(xs, i);
 
-    if (elt == R_NilValue) {
+    if (x == R_NilValue) {
       continue;
     }
 
-    R_len_t index_size = vec_size(VECTOR_ELT(indices, i));
+    R_len_t index_size = Rf_length(VECTOR_ELT(indices, i));
     out_size += index_size;
 
-    if (has_names) {
-      p_sizes[i] = index_size;
-    }
-
-    // Each element of `x` is recycled to its corresponding index's size
-    elt = vec_recycle(elt, index_size, args_empty);
-    SET_VECTOR_ELT(x, i, elt);
+    // Each element of `xs` is recycled to its corresponding index's size
+    x = vec_recycle(x, index_size, args_empty);
+    SET_VECTOR_ELT(xs, i, x);
   }
 
-  indices = PROTECT(vec_as_indices(indices, out_size, R_NilValue));
+  SEXP locs = PROTECT(vec_as_indices(indices, out_size, R_NilValue));
 
-  PROTECT_INDEX proxy_pi;
   SEXP proxy = vec_proxy(ptype);
+  PROTECT_INDEX proxy_pi;
   PROTECT_WITH_INDEX(proxy, &proxy_pi);
+
   proxy = vec_init(proxy, out_size);
   REPROTECT(proxy, proxy_pi);
 
+  SEXP out_names = R_NilValue;
   PROTECT_INDEX out_names_pi;
-  SEXP out_names = vctrs_shared_empty_chr;
-  if (has_names) {
-    out_names = Rf_allocVector(STRSXP, out_size);
-  }
   PROTECT_WITH_INDEX(out_names, &out_names_pi);
 
   const struct vec_assign_opts unchop_assign_opts = {
     .assign_names = assign_names
   };
 
   for (R_len_t i = 0; i < x_size; ++i) {
-    SEXP elt = VECTOR_ELT(x, i);
+    SEXP x = VECTOR_ELT(xs, i);
 
-    if (elt == R_NilValue) {
+    if (x == R_NilValue) {
       continue;
     }
 
-    SEXP index = VECTOR_ELT(indices, i);
+    SEXP loc = VECTOR_ELT(locs, i);
 
     // Total ownership of `proxy` because it was freshly created with `vec_init()`
-    proxy = vec_proxy_assign_opts(proxy, index, elt, VCTRS_OWNED_true, &unchop_assign_opts);
+    proxy = vec_proxy_assign_opts(proxy, loc, x, VCTRS_OWNED_true, &unchop_assign_opts);
     REPROTECT(proxy, proxy_pi);
 
-    if (has_names) {
-      R_len_t size = p_sizes[i];
-      SEXP outer = (has_outer_names) ? STRING_ELT(x_names, i) : R_NilValue;
-      SEXP inner = PROTECT(vec_names(elt));
-      SEXP elt_names = PROTECT(apply_name_spec(name_spec, outer, inner, size));
-      if (elt_names != R_NilValue) {
-        out_names = chr_assign(out_names, index, elt_names, VCTRS_OWNED_true);
-        REPROTECT(out_names, out_names_pi);
+    if (assign_names) {
+      R_len_t size = Rf_length(loc);
+      SEXP outer = xs_is_named ? STRING_ELT(xs_names, i) : R_NilValue;
+      SEXP inner = PROTECT(vec_names(x));
+      SEXP x_nms = PROTECT(apply_name_spec(name_spec, outer, inner, size));
+
+      if (x_nms != R_NilValue) {
+        R_LAZY_ALLOC(out_names, out_names_pi, STRSXP, out_size);
+        if (x_nms != chrs_empty) {
+          out_names = chr_assign(out_names, loc, x_nms, VCTRS_OWNED_true);
+          REPROTECT(out_names, out_names_pi);
+        }
```

Comment:
What is this if check for? Can `chr_assign()` not handle this or is it special behavior?

Reply:
I'll add a comment that this is to prevent recycling. Probably can be removed once we support ALTREP repetitions in recycling and assignment.

Reply:
Even with altrep we'll be better off with this check to avoid the assignment loop.